src/coordinator.rs

```diff
@@ -244,78 +266,157 @@ impl<'roc> Builder<'roc> {
     }
 }
 
+type DoneMsg = (job::Key<job::Base>, Option<Workspace>);
+
 #[derive(Debug)]
-pub struct Coordinator<'roc> {
+pub struct Coordinator {
     store: Store,
+    runner_builder: RunnerBuilder,
 
     roots: Vec<job::Key<job::Base>>,
+    max_local_jobs: usize,
 
     // caches
     path_to_hash: HashMap<PathBuf, String>,
+    final_keys: HashMap<job::Key<job::Base>, job::Key<job::Final>>,
 
     // note:  this mapping is only safe to use in the context of a single
     // execution since a job's final key may change without the base key
     // changing. Practically speaking, this just means you shouldn't store it!
     job_to_content_hash: HashMap<job::Key<job::Base>, store::Item>,
 
     // which jobs should run when?
-    jobs: HashMap<job::Key<job::Base>, Job<'roc>>,
+    jobs: HashMap<job::Key<job::Base>, Job>,
     blocked: HashMap<job::Key<job::Base>, HashSet<job::Key<job::Base>>>,
+
+    // what's the state of the coordinator while running?
     ready: Vec<job::Key<job::Base>>,
+    running: FuturesUnordered<JoinHandle<Result<DoneMsg>>>,
 }
 
-impl<'roc> Coordinator<'roc> {
-    pub fn has_outstanding_work(&self) -> bool {
-        !self.blocked.is_empty() || !self.ready.is_empty()
+impl<'roc> Coordinator {
+    /// Run the build from start to finish.
+    pub async fn run(&mut self) -> Result<()> {
+        log::trace!("scheduling immediately-available jobs");
+        self.schedule()
+            .await
+            .context("could not start immediately-ready jobs")?;
+
+        log::trace!("starting coordinator loop");
+        while let Some(join_res) = self.running.next().await {
+            match join_res {
+                Ok(Ok(done_msg)) => self
+                    .handle_done(done_msg)
+                    .await
+                    .context("could not finish job")?,
+                Ok(Err(err)) => todo!("handle jobs that return in failure. Error was: #{err:#?}"),
+                Err(err) => todo!("handle jobs that cannot be joined. Error was: #{err:#?}"),
+            }
+        }
+
+        Ok(())
     }
 
-    pub fn run_next<R: Runner>(&mut self, runner: &R) -> Result<()> {
-        let id = match self.ready.pop() {
-            Some(id) => id,
-            None => anyhow::bail!("no work ready to do"),
-        };
+    /// Start any outstanding work according to our scheduling rules. Right
+    /// now that just means that we won't ever be running more jobs than
+    /// `self.max_local_jobs`.
+    async fn schedule(&mut self) -> Result<()> {
+        let maximum_schedulable = (self.max_local_jobs - self.running.len()).max(0);
```

Comment:
```suggestion
        let maximum_schedulable = self.max_local_jobs.saturating_sub(self.running.len());
```

https://doc.rust-lang.org/std/primitive.usize.html#method.saturating_sub

Reply:
Also possibly useful:

```rust
debug_assert!(self.running.len() <= self.max_local_jobs);
```